### PR TITLE
Completed parser and generator

### PR DESCRIPTION
**Pull request by kwxm**

### CODE REVIEW

File: field/TinyLang/Field/Core.hs [hunk 1]
```diff
@@ -201,6 +203,7 @@ withGeqUni _     _     _ z = z
 withGeqUnOp :: UnOp f a1 b1 -> UnOp f a2 b2 -> ((a1 ~ a2, b1 ~ b2) => d) -> d -> d
 withGeqUnOp Not  Not  y _ = y
 withGeqUnOp Neq0 Neq0 y _ = y
+withGeqUnOp Neg  Neg  y _ = y
```
Comment (effectfully) [inline field/TinyLang/Field/Core.hs:206]: Thanks again for catching this.

File: field/TinyLang/Field/Printer.hs [hunk 2]
```diff
@@ -60,3 +60,6 @@ toStringNoIDs = toString NoIDs
 -- | Convert an Expr to a String, including Unique IDs in variable names
 toStringWithIDs :: Show f => Expr f a -> String
 toStringWithIDs = toString WithIDs
+
+tos :: (Show f) => SomeUniExpr f -> String
+tos (SomeUniExpr _ e) = toStringNoIDs e
```
Comment (effectfully) [inline field/TinyLang/Field/Printer.hs]: Why not something like `someToStringWithIDs`? `tos` is a weird name for an exported top-level function.

File: field/TinyLang/Field/Printer.hs [hunk 3]
```diff
@@ -34,7 +34,7 @@ toStringBinOp Div = " / "
 
 -- Do we want () round something when printing it inside some other expression?
 isSimple :: Expr f a -> Bool
-isSimple (EVal _) = True
+-- isSimple (EVal _) = True  -- No!  Rationals cause problems: for example, neq0 (-1) % 2 is hard to parse
```
Comment (effectfully) [inline field/TinyLang/Field/Printer.hs:38]: That's unfortunate. I guess we need a class that determines whether `f` is simple or not. I'll do something like that in future together with adding a precedence- and associativity-aware pretty-printer.

File: field/TinyLang/Field/F17.hs [hunk 4]
```diff
@@ -48,6 +48,7 @@ instance Field F17 where
         _  -> error $ "F17 is not in the [0, 16] range: " ++ show n
 
 -- | Choose a random element uniformly. If you use this in
+
 -- randomly-generated expressions there's a good chance that you'll
 -- get division by zero (-> undefined).
```
Comment (effectfully) [inline field/TinyLang/Field/F17.hs]: Do we need the new newline or is it a typo?

**Review by effectfully — APPROVED** [review]
Perfect, thank you. I only have a few nitpicks.

File: field/TinyLang/Field/F17.hs (hunk 4 above)
Comment (kwxm) [reply field/TinyLang/Field/F17.hs]: Typo

File: field/TinyLang/Field/Printer.hs (hunk 3 above)
Comment (kwxm) [reply field/TinyLang/Field/Printer.hs:38]: I suppose it doesn't really matter.  In the previous version this made things look a bit nicer, but this version's a bit more complicated so it's just easier to put () round everything.  As you say, a proper pretty-printer might be worthwhile if you need it.

File: field/TinyLang/Field/Printer.hs (hunk 2 above)
Comment (kwxm) [reply field/TinyLang/Field/Printer.hs]: Sorry!  I stuck that in for experiments and then forgot to make it more sensible.

**Comment by kwxm** [comment]
I'll fix these later on.

File: field/TinyLang/Field/F17.hs (hunk 4 above)
Comment (kwxm) [reply field/TinyLang/Field/F17.hs]: Fixed

File: field/TinyLang/Field/Printer.hs (hunk 2 above)
Comment (kwxm) [reply field/TinyLang/Field/Printer.hs]: I've now made this `someExprToString` and exposed the `WithIDs | NoIDs` type in the module interface instead of having separate `exprToStringNoIDs` and `exprToStringWithIDs` functions.

**Comment by effectfully** [comment]
Thank you!
